toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -222,3 +233,93 @@ func customizeImageHelper(buildDir string, baseConfigPath string, config *imagec
 
 	return nil
 }
+
+func customizeVerityImageHelper(buildDir string, baseConfigPath string, config *imagecustomizerapi.Config,
+	buildImageFile string, rpmsSources []string, useBaseImageRpmRepos bool,
+) error {
+	var err error
+
+	// Connect the disk image to an NBD device using qemu-nbd
+	// Find a free NBD device
+	nbdDevice, err := findFreeNBDDevice()
+	if err != nil {
+		return fmt.Errorf("failed to find a free nbd device: %v", err)
+	}
+
+	err = shell.ExecuteLiveWithErr(1, "qemu-nbd", "-c", nbdDevice, "-f", "raw", buildImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to connect nbd %s to image %s: %s", nbdDevice, buildImageFile, err)
+	}
+	defer func() {
+		// Disconnect the NBD device when the function returns
+		err = shell.ExecuteLiveWithErr(1, "qemu-nbd", "-d", nbdDevice)
+		if err != nil {
+			return
+		}
+	}()
+
+	diskPartitions, err := diskutils.GetDiskPartitions(nbdDevice)
+	if err != nil {
+		return err
+	}
+
+	// Extract the partition block device path.
+	dataPartition, err := idToPartitionBlockDevicePath(config.SystemConfig.Verity.DataPartition.IdType, config.SystemConfig.Verity.DataPartition.Id, nbdDevice, diskPartitions)
+	if err != nil {
+		return err
+	}
+	hashPartition, err := idToPartitionBlockDevicePath(config.SystemConfig.Verity.HashPartition.IdType, config.SystemConfig.Verity.HashPartition.Id, nbdDevice, diskPartitions)
+	if err != nil {
+		return err
+	}
+
+	// Extract root hash using regular expressions.
+	verityOutput, _, err := shell.Execute("veritysetup", "format", dataPartition, hashPartition)
+	if err != nil {
+		return fmt.Errorf("failed to calculate root hash:\n%w", err)
+	}
+
+	var rootHash string
+	rootHashRegex := regexp.MustCompile(`Root hash:\s+([0-9a-fA-F]+)`)
+
+	rootHashMatches := rootHashRegex.FindStringSubmatch(verityOutput)
+	if len(rootHashMatches) <= 1 {
+		return fmt.Errorf("failed to parse root hash from veritysetup output")
+	}
+	rootHash = rootHashMatches[1]
+
+	bootPartitionTmpDir := filepath.Join(buildDir, tmpParitionDirName)
+	for _, diskPartition := range diskPartitions {
+		switch diskPartition.FileSystemType {
+		case "ext4", "vfat", "xfs":
+
+		default:
+			// Skips file system types that aren't known to support the boot loader partition.
+			continue
+		}
+
+		// Temporarily mount the partition.
+		partitionMount, err := safemount.NewMount(diskPartition.Path, bootPartitionTmpDir, diskPartition.FileSystemType, 0, "", true)
+		if err != nil {
+			return fmt.Errorf("failed to mount partition (%s):\n%w", diskPartition.Path, err)
+		}
+		defer partitionMount.Close()
```

Comment:
You need to also call `partitionMount.CleanClose()` and check its error result.

`.Close()` runs asynchronously and doesn't guarantee that the mount closed successfully. This is fine when there is already an error. But you don't want the tool to report a success if there were problems unmounting a partition.